File: python/deepstrike/agent.py
"""Public executable Agent contract.

This surface describes an agent before a host chooses a provider, execution plane, or Kernel run.
The declaration stays provider-neutral; execution resolves a provider through the explicit
``runtime_binding`` supplied by the host.
"""
from __future__ import annotations

from dataclasses import dataclass
import uuid
import time
from collections.abc import AsyncIterator
from typing import Any, Literal, Mapping, Sequence, TypeAlias

from deepstrike.types.agent import AgentCapabilityFilter


ModelRef: TypeAlias = str | dict[str, Any]
AgentDefinition: TypeAlias = Mapping[str, Any]
AgentMemory: TypeAlias = Any


@dataclass(frozen=True)
class MemoryReference:
    """Serializable durable-memory binding for an agent declared before a host store exists."""

    namespace: str | None = None
    kind: Literal["durable"] = "durable"


class RunResult(dict[str, Any]):
    """Completed run data with both mapping and attribute-style access.

    Mapping access keeps the 0.2.x facade compatible while attributes make the
    result pleasant to use in normal Python code.
    """

    def __init__(
        self,
        *,
        output: str,
        run_id: str | None,
        session_id: str,
        status: str,
        usage: dict[str, Any] | None = None,
        evidence: list[dict[str, Any]] | None = None,
    ) -> None:
        super().__init__(
            output=output,
            run_id=run_id,
            session_id=session_id,
            status=status,
            usage=usage,
            evidence=evidence or [],
        )

    output: str
    run_id: str | None
    session_id: str
    status: str
    usage: dict[str, Any] | None
    evidence: list[dict[str, Any]]

    def __getattr__(self, name: str) -> Any:
        try:
            return self[name]
        except KeyError as exc:
            raise AttributeError(name) from exc


class AgentSession:
    """A reusable, resumable session owned by one :class:`Agent`."""

    def __init__(self, agent: "Agent", session_id: str) -> None:
        self.agent = agent
        self.id = session_id
        self._runner = agent._runner_for(session_id)

    async def stream(
        self,
        goal: str,
        *,
        max_turns: int | None = None,
    ) -> AsyncIterator[Any]:
        runner = self._runner if max_turns is None else self.agent._runner_for(self.id, max_turns=max_turns)
        async for event in runner.run(goal=goal, session_id=self.id):
            yield event

    async def resume(self) -> AsyncIterator[Any]:
        async for event in self._runner.wake(self.id):
            yield event

    def interrupt(self, reason: str = "user") -> None:
        self._runner.interrupt(reason)  # type: ignore[arg-type]

    async def run(self, goal: str, *, max_turns: int | None = None) -> RunResult:
        return await self.agent._run_in_session(self, goal, max_turns=max_turns)


class Agent:
    """Provider-neutral declaration and executable public Agent handle.

    ``tools`` may contain executable ``RegisteredTool`` instances or JSON-safe tool descriptors.
    The latter carry schema only and do not create executable capabilities.
    """

    def __init__(
        self,
        name: str,
        *,
        description: str | None = None,
        instructions: str | None = None,
        model: ModelRef | None = None,
        capability_filter: AgentCapabilityFilter | Mapping[str, Any] | None = None,
        tools: Sequence[Any] | None = None,
        mcp_servers: Sequence[Mapping[str, Any]] | None = None,
        skills: Sequence[Mapping[str, Any]] | None = None,
        memory: AgentMemory | None = None,
        knowledge: Sequence[Mapping[str, Any]] | None = None,
        handoffs: Sequence[Mapping[str, Any]] | None = None,
        provider_options: Mapping[str, Any] | None = None,
        output_schema: Mapping[str, Any] | None = None,
        metadata: Mapping[str, Any] | None = None,
        guardrails: Sequence[Mapping[str, Any]] | None = None,
        runtime_binding: Mapping[str, Any] | None = None,
    ) -> None:
        if not name:
            raise ValueError("agent name is required")
        self.name = name
        self.description = description
        self.instructions = instructions
        self.model = model
        self.capability_filter = capability_filter
        self.tools = list(tools) if tools is not None else None
        self.mcp_servers = list(mcp_servers) if mcp_servers is not None else None
        self.skills = list(skills) if skills is not None else None
        self.memory = memory
        self.knowledge = list(knowledge) if knowledge is not None else None
        self.handoffs = list(handoffs) if handoffs is not None else None
        self.provider_options = dict(provider_options) if provider_options is not None else None
        self.output_schema = dict(output_schema) if output_schema is not None else None
        self.metadata = dict(metadata) if metadata is not None else None
        self.guardrails = list(guardrails) if guardrails is not None else None
        self.runtime_binding = dict(runtime_binding) if runtime_binding is not None else None
        self._session_log = None
        self._runners: dict[str, Any] = {}

    @property
    def declaration(self) -> dict[str, Any]:
        """Return the serializable agent declaration used by runtime adapters."""
        return {
            "name": self.name,
            "description": self.description,
            "instructions": self.instructions,
            "model": self.model,
            "capability_filter": self.capability_filter,
            "tools": self.tools,
            "mcp_servers": self.mcp_servers,
            "skills": self.skills,
            "memory": self.memory,
            "knowledge": self.knowledge,
            "handoffs": self.handoffs,
            "provider_options": self.provider_options,
            "output_schema": self.output_schema,
            "metadata": self.metadata,
            "guardrails": self.guardrails,
        }

    def session(self, session_id: str | None = None) -> AgentSession:
        """Return a stable session handle, creating an id when omitted."""
        return AgentSession(self, session_id or f"agent-{uuid.uuid4()}")

    def _provider(self) -> Any:
        if not self.runtime_binding:
            raise RuntimeError(f'agent "{self.name}" has no runtime binding')
        provider = self.runtime_binding.get("provider")
        if provider is None:
            provider_for = self.runtime_binding.get("provider_for")
            provider = provider_for(self.model) if callable(provider_for) else None
        if provider is None:
            raise RuntimeError(f'agent "{self.name}" has no runtime provider binding')
        return provider

    def _runner_for(self, session_id: str, *, max_turns: int | None = None) -> Any:
        runner = self._runners.get(session_id)
        if runner is not None and max_turns is None:
            return runner
        from deepstrike.runtime.execution_plane import LocalExecutionPlane
        from deepstrike.runtime.runner import RuntimeOptions, RuntimeRunner
        from deepstrike.runtime.session_log import InMemorySessionLog

        if self._session_log is None:
            self._session_log = self.runtime_binding.get("session_log") if self.runtime_binding else None
            if self._session_log is None:
                self._session_log = InMemorySessionLog()
        raw_options = dict((self.runtime_binding or {}).get("runtime_options") or {})
        raw_options.pop("provider", None)
        raw_options.pop("session_log", None)
        raw_options.pop("execution_plane", None)
        raw_options.pop("agent_id", None)
        raw_options.pop("system_prompt", None)
        if max_turns is not None:
            raw_options["max_turns"] = max_turns
        binding = self.runtime_binding or {}
        if binding.get("memory_store") is not None:
            raw_options.setdefault("memory_store", binding["memory_store"])
        if binding.get("memory_scope") is not None:
            raw_options.setdefault("memory_scope", binding["memory_scope"])
        configured_plane = (self.runtime_binding or {}).get("execution_plane")
        plane = configured_plane or LocalExecutionPlane()
        if configured_plane is None:
            executable_tools = [tool for tool in (self.tools or []) if hasattr(tool, "schema")]
            if executable_tools:
                plane.register(*executable_tools)
        options = RuntimeOptions(
            provider=self._provider(),
            execution_plane=plane,
            session_log=self._session_log,
            agent_id=self.name,
            **({"system_prompt": self.instructions} if self.instructions else {}),
            **raw_options,
        )
        runner = RuntimeRunner(options)
        if max_turns is None:
            self._runners[session_id] = runner
        return runner

    def _memory_binding(self) -> tuple[Any, Any]:
        binding = self.runtime_binding or {}
        store = binding.get("memory_store")
        scope = binding.get("memory_scope")
        if store is None or scope is None:
            raise RuntimeError(
                f'agent "{self.name}" memory APIs require runtime_binding.memory_store and memory_scope'
            )
        return store, scope

    async def remember(
        self,
        memory: Any,
        *,
        session_id: str | None = None,
        name: str | None = None,
        kind: str = "reference",
        description: str = "",
        pinned: bool = False,
    ) -> Any:
        """Persist one durable memory through the same runner/kernel path as a live run."""
        from deepstrike.memory.protocols import MemoryProvenance, MemoryRecord

        _store, scope = self._memory_binding()
        if isinstance(memory, MemoryRecord):
            record = memory
        else:
            content = str(memory)
            now = int(time.time() * 1000)
            record = MemoryRecord(
                record_id=f"mem-{uuid.uuid4()}",
                scope=scope,
                name=name or "memory",
                kind=kind,  # type: ignore[arg-type]
                content=content,
                description=description,
                provenance=MemoryProvenance(author="host", trust="user_asserted", session_id=session_id),
                created_at=now,
                updated_at=now,
                pinned=pinned,
            )
        sid = session_id or f"agent-memory-{uuid.uuid4()}"
        runner = self._runner_for(sid)
        await runner.write_memory(record, session_id=sid, agent_id=self.name)
        return record

    async def recall(
        self,
        query: str,
        *,
        session_id: str | None = None,
        top_k: int = 5,
        min_score: float | None = None,
    ) -> list[Any]:
        """Search durable memory using the configured store and scope."""
        _store, scope = self._memory_binding()
        from deepstrike.memory.protocols import MemoryQuery

        sid = session_id or f"agent-memory-{uuid.uuid4()}"
        runner = self._runner_for(sid)
        return await runner.query_memory(
            MemoryQuery(scope=scope, query=query, top_k=top_k, min_score=min_score),
            session_id=sid,
            agent_id=self.name,
        )

    async def workflow(self, spec: Any, *, session_id: str | None = None) -> Any:
        """Run a declarative workflow through the runner's governed workflow path."""
        sid = session_id or f"agent-workflow-{uuid.uuid4()}"
        return await self._runner_for(sid).run_workflow(spec, session_id=sid)

    async def _run_in_session(self, session: AgentSession, goal: str, *, max_turns: int | None = None) -> RunResult:
        from deepstrike.providers.stream import DoneEvent, ErrorEvent, TextDelta, UsageEvent

        output: list[str] = []
        usage: dict[str, Any] | None = None
        status = "completed"
        async for event in session.stream(goal, max_turns=max_turns):
            if isinstance(event, TextDelta):
                output.append(event.delta)
            elif isinstance(event, UsageEvent):
                usage = {
                    "input_tokens": event.input_tokens,
                    "output_tokens": event.output_tokens,
                    "total_tokens": event.total_tokens,
                }
            elif isinstance(event, DoneEvent):
                status = event.status
            elif isinstance(event, ErrorEvent):
                status = "error"
        entries = await self._session_log.read(session.id) if self._session_log is not None else []
        started = next((entry.event for entry in entries if entry.event.get("kind") == "run_started"), None)
        evidence = [entry.event for entry in entries if entry.event.get("kind") in {"provider_attempt", "llm_completed"}]
        return RunResult(
            output="".join(output),
            run_id=started.get("run_id") if started else None,
            session_id=session.id,
            status=status,
            usage=usage,
            evidence=evidence,
        )

    async def run(self, goal: str, *, session_id: str | None = None, max_turns: int | None = None) -> RunResult:
        """Execute one goal and return a reusable, structured result."""
        session = self.session(session_id)
        return await self._run_in_session(session, goal, max_turns=max_turns)

    async def stream(self, goal: str, *, session_id: str | None = None, max_turns: int | None = None):
        """Stream host events while retaining the session for later resume."""
        session = self.session(session_id)
        async for event in session.stream(goal, max_turns=max_turns):
            yield event

    def close(self) -> None:
        """Interrupt active runs owned by this agent."""
        for runner in self._runners.values():
            runner.interrupt("host_shutdown")


def create_agent(name: str, **kwargs: Any) -> Agent:
    """Create the executable public Agent handle."""
    return Agent(name, **kwargs)

File: wasm/src/agent.ts
import type { AgentCapabilityFilter, WorkflowOutcome, WorkflowSpec } from "./runtime/types/agent.js"
import type { Memory, WorkingMemory, MemoryRecord, MemoryRecall, MemoryKind } from "./memory/index.js"
import type { RegisteredTool } from "./tools/index.js"
import type { LLMProvider, StreamEvent, ContentPart, UsageEvent } from "./types.js"
import { RuntimeRunner } from "./runtime/runner.js"
import { LocalExecutionPlane, type ExecutionPlane } from "./runtime/execution-plane.js"
import { InMemorySessionLog, type SessionEvent, type SessionLog } from "./runtime/session-log.js"
import { extractJsonValue, validateAgainstSchema } from "./runtime/output-schema.js"
import type { SignalSource } from "./signals/index.js"
import type { McpExecutionPlane } from "./runtime/mcp-transport.js"

type JsonSchema = Record<string, unknown>
export interface MemoryReference { kind?: "durable"; namespace?: string }
export type AgentMemory = Memory | WorkingMemory | MemoryReference
export type ModelRef = string | ModelRequirement
export interface ModelRequirement { capability?: { reasoning?: boolean; vision?: boolean; toolUse?: boolean }; contextWindow?: number; latencyClass?: string; costClass?: string }
export interface AgentToolDefinition { name: string; description?: string; parameters?: Record<string, unknown>; providerOptions?: Record<string, unknown> }
export type McpTransport = { kind: "stdio"; command: string; args?: string[] } | { kind: "http" | "sse"; url: string } | { kind: "custom"; [key: string]: unknown }
export interface MCPServer { name?: string; transport: McpTransport; tools?: string[]; resources?: boolean; prompts?: boolean; auth?: Record<string, unknown>; metadata?: Record<string, unknown>; providerOptions?: Record<string, unknown> }
export interface Skill { name: string; description?: string; instructions?: string; resources?: unknown[]; scripts?: unknown[]; tools?: unknown[]; mcpServers?: unknown[]; knowledge?: unknown[]; metadata?: Record<string, unknown>; providerOptions?: Record<string, unknown> }
export type KnowledgeSourceRef = { kind: "file" | "directory" | "text" | "url" | "vector" | "custom"; [key: string]: unknown }
export interface Knowledge { id?: string; name?: string; source: KnowledgeSourceRef; description?: string; metadata?: Record<string, unknown>; providerOptions?: Record<string, unknown> }
export type AgentRef = string | { name: string }
export interface Handoff { agent: AgentRef; description?: string; inputSchema?: JsonSchema; metadata?: Record<string, unknown>; providerOptions?: Record<string, unknown> }
export interface Guardrail { name: string; description?: string; metadata?: Record<string, unknown> }

export interface AgentRuntimeBinding {
  provider?: LLMProvider
  providerFor?: (model: string) => LLMProvider | undefined
  runtimeOptions?: Partial<import("./runtime/runner.js").RuntimeOptions>
  sessionLog?: SessionLog
  executionPlane?: ExecutionPlane
  mcpPlane?: McpExecutionPlane
  resolveAgent?: (name: string) => Agent | undefined | Promise<Agent | undefined>
  signalSource?: SignalSource
}
export interface AgentOptions {
  name: string; description?: string; instructions?: string; model?: ModelRef
  capabilityFilter?: AgentCapabilityFilter; tools?: Array<RegisteredTool | AgentToolDefinition>
  mcpServers?: MCPServer[]; skills?: Skill[]; memory?: AgentMemory; knowledge?: Knowledge[]
  handoffs?: Handoff[]; providerOptions?: Record<string, unknown>; outputSchema?: JsonSchema
  metadata?: Record<string, unknown>; guardrails?: Guardrail[]; runtimeBinding?: AgentRuntimeBinding
}
export interface AgentRunOptions {
  sessionId?: string; maxTurns?: number; maxTotalTokens?: number; timeoutMs?: number
  metadata?: Record<string, unknown>; providerOptions?: Record<string, unknown>
  attachments?: ContentPart[]; cancelSignal?: AbortSignal
}
export interface AgentRunResult {
  output: string; runId?: string; sessionId: string
  status: "completed" | "partial" | "failed" | "cancelled"
  usage?: { inputTokens: number; outputTokens: number; totalTokens: number }
  outputValidation?: { ok: boolean; errors: string[] }
  evidence?: { route?: unknown; measurement?: unknown; contextBinding?: unknown }
}
export interface MemoryInput { name: string; content: string; description?: string; kind?: MemoryKind; confidence?: number; pinned?: boolean; ttlDays?: number }
export interface RecallOptions { topK?: number; kinds?: MemoryKind[]; minScore?: number }
export interface DelegationResult { output: string; status: "completed" | "partial" | "failed" }

function statusOf(reason: string): AgentRunResult["status"] {
  if (["completed", "success", "done"].includes(reason)) return "completed"
  if (["cancelled", "user", "user_abort", "deadline", "timeout"].includes(reason)) return "cancelled"
  if (["failed", "error", "invalid_arg"].includes(reason)) return "failed"
  return "partial"
}

export class AgentSession {
  readonly id: string
  private activeRunner?: RuntimeRunner
  constructor(private readonly owner: Agent, id: string) { this.id = id }
  history(fromSeq = 0) { return this.owner.sessionLog.read(this.id, fromSeq) }
  latestSeq() { return this.owner.sessionLog.latestSeq(this.id) }
  async workflowTrace() {
    const kinds = new Set(["workflow_batch_spawned", "workflow_node_completed", "workflow_nodes_submitted", "workflow_completed"])
    return (await this.history()).filter(entry => kinds.has(entry.event.kind))
  }
  async workflowReplay() {
    const events = await this.workflowTrace()
    const outcomes = new Map<string, Record<string, unknown>>()
    let completed = false
    for (const entry of events) {
      if (entry.event.kind === "workflow_node_completed") {
        const id = entry.event.agent_id
        if (id) outcomes.set(id, entry.event as unknown as Record<string, unknown>)
      }
      if (entry.event.kind === "workflow_completed") {
        completed = true
        for (const outcome of entry.event.node_outcomes) {
          const id = outcome.node_id
          if (id) outcomes.set(id, outcome as unknown as Record<string, unknown>)
        }
      }
    }
    return { events, nodeOutcomes: [...outcomes.values()], completed }
  }
  private runner(goal: string, options: AgentRunOptions = {}) { return this.owner.createRunner(goal, this.id, options) }

  async *stream(goal: string, options: Omit<AgentRunOptions, "sessionId"> = {}): AsyncIterable<StreamEvent> {
    const runner = await this.runner(goal, { ...options, sessionId: this.id })
    this.activeRunner = runner
    const abort = () => runner.interrupt("user")
    if (options.cancelSignal?.aborted) abort()
    else options.cancelSignal?.addEventListener("abort", abort, { once: true })
    try { yield* runner.run({ sessionId: this.id, goal, attachments: options.attachments, extensions: options.providerOptions }) }
    finally { options.cancelSignal?.removeEventListener("abort", abort); this.activeRunner = undefined }
  }
  async run(goal: string, options: Omit<AgentRunOptions, "sessionId"> = {}) {
    const events: StreamEvent[] = []
    for await (const event of this.stream(goal, options)) events.push(event)
    return this.owner.result(this.id, events)
  }
  async *resume(options: Omit<AgentRunOptions, "sessionId"> = {}): AsyncIterable<StreamEvent> {
    const runner = await this.runner("resume", { ...options, sessionId: this.id })
    this.activeRunner = runner
    try { yield* runner.wake(this.id, options.providerOptions) }
    finally { this.activeRunner = undefined }
  }
  interrupt(reason: "user" | "deadline" | "lease_lost" | "host_shutdown" = "user") { this.activeRunner?.interrupt(reason) }
  remember(input: MemoryInput): Promise<MemoryRecord> { return this.owner.remember(input, this.id) }
  recall(query: string, options: RecallOptions = {}): Promise<MemoryRecall[]> { return this.owner.recall(query, options, this.id) }
  async workflow(spec: WorkflowSpec): Promise<WorkflowOutcome> {
    const runner = await this.runner("workflow", { sessionId: this.id }); this.activeRunner = runner
    try { return await runner.runWorkflow(spec, { sessionId: this.id }) } finally { this.activeRunner = undefined }
  }
}

export class Agent {
  readonly name: string; readonly description?: string; readonly instructions?: string; readonly model?: ModelRef
  readonly capabilityFilter?: AgentCapabilityFilter; readonly tools?: Array<RegisteredTool | AgentToolDefinition>
  readonly mcpServers?: MCPServer[]; readonly skills?: Skill[]; readonly memory?: AgentMemory; readonly knowledge?: Knowledge[]
  readonly handoffs?: Handoff[]; readonly providerOptions?: Record<string, unknown>; readonly outputSchema?: JsonSchema
  readonly metadata?: Record<string, unknown>; readonly guardrails?: Guardrail[]; readonly runtimeBinding?: AgentRuntimeBinding
  readonly sessionLog: SessionLog
  private readonly sessions = new Map<string, AgentSession>()
  constructor(options: AgentOptions) {
    this.name = options.name; this.description = options.description; this.instructions = options.instructions; this.model = options.model
    this.capabilityFilter = options.capabilityFilter; this.tools = options.tools; this.mcpServers = options.mcpServers
    this.skills = options.skills; this.memory = options.memory; this.knowledge = options.knowledge; this.handoffs = options.handoffs
    this.providerOptions = options.providerOptions; this.outputSchema = options.outputSchema; this.metadata = options.metadata; this.guardrails = options.guardrails
    this.runtimeBinding = options.runtimeBinding; this.sessionLog = options.runtimeBinding?.sessionLog ?? options.runtimeBinding?.runtimeOptions?.sessionLog ?? new InMemorySessionLog()
  }
  session(id = `session-${crypto.randomUUID()}`) { if (!this.sessions.has(id)) this.sessions.set(id, new AgentSession(this, id)); return this.sessions.get(id)! }
  async createRunner(_goal: string, _sessionId: string, options: AgentRunOptions = {}) {
    const binding = this.runtimeBinding
    const provider = binding?.provider ?? (typeof this.model === "string" ? binding?.providerFor?.(this.model) : undefined)
    if (!provider) throw new Error(`agent "${this.name}" has no runtime provider binding`)
    const base = { ...(binding?.runtimeOptions ?? {}) } as Record<string, unknown>
    if (options.maxTurns !== undefined) base.maxTurns = options.maxTurns
    if (options.maxTotalTokens !== undefined) base.maxTotalTokens = options.maxTotalTokens
    if (options.timeoutMs !== undefined) base.timeoutMs = options.timeoutMs
    if (options.providerOptions) base.extensions = { ...((base.extensions as Record<string, unknown> | undefined) ?? {}), ...options.providerOptions }
    const mcpPlane = binding?.mcpPlane
    if (mcpPlane) await mcpPlane.connect()
    return new RuntimeRunner({ provider, executionPlane: binding?.executionPlane ?? mcpPlane ?? new LocalExecutionPlane(), sessionLog: this.sessionLog, maxTokens: 32_000, agentId: this.name, ...(this.instructions ? { systemPrompt: this.instructions } : {}), ...base } as import("./runtime/runner.js").RuntimeOptions)
  }
  async run(goal: string, options: AgentRunOptions = {}) { return this.session(options.sessionId).run(goal, options) }
  stream(goal: string, options: AgentRunOptions = {}) { return this.session(options.sessionId).stream(goal, options) }
  resume(sessionId: string, options: Omit<AgentRunOptions, "sessionId"> = {}) { return this.session(sessionId).resume(options) }
  interrupt(reason: "user" | "deadline" | "lease_lost" | "host_shutdown" = "user", sessionId?: string) { sessionId ? this.session(sessionId).interrupt(reason) : [...this.sessions.values()].forEach(s => s.interrupt(reason)) }
  workflow(spec: WorkflowSpec, options: { sessionId?: string } = {}) { return this.session(options.sessionId).workflow(spec) }
  async remember(input: MemoryInput, sessionId?: string): Promise<MemoryRecord> {
    if (!this.memory || typeof (this.memory as Memory).put !== "function") throw new Error(`agent "${this.name}" memory is not runtime-bound`)
    const now = Date.now()
    const record: MemoryRecord = {
      record_id: crypto.randomUUID(), scope: (this.memory as Memory & { scope?: { tenant_id: string; namespace: string } }).scope
        ?? { tenant_id: "default", namespace: (this.memory as Memory).namespace ?? this.name },
      name: input.name, kind: input.kind ?? "reference", content: input.content,
      description: input.description ?? input.name,
      provenance: { author: "host", trust: "user_asserted", evidence_refs: [], session_id: sessionId ?? `memory-${crypto.randomUUID()}` },
      created_at: now, updated_at: now, recall_count: 0, confidence: input.confidence ?? 1,
      links: [], pinned: input.pinned ?? false, ...(input.ttlDays === undefined ? {} : { ttl_days: input.ttlDays }),
    }
    await (this.memory as Memory).put(record)
    return record
  }
  async recall(query: string, options: RecallOptions = {}, _sessionId?: string): Promise<MemoryRecall[]> {
    if (!this.memory || typeof (this.memory as Memory).search !== "function") throw new Error(`agent "${this.name}" memory is not runtime-bound`)
    const records = await (this.memory as Memory).search(query, { topK: options.topK ?? 8, kinds: options.kinds, minScore: options.minScore })
    return records.map(record => ({ record, score: 1, why: "memory facade search" }))
  }
  async delegate(request: { target: AgentRef; goal: string; metadata?: Record<string, unknown>; providerOptions?: Record<string, unknown> }): Promise<DelegationResult> {
    const targetName = typeof request.target === "string" ? request.target : request.target.name
    const declared = (this.handoffs ?? []).some(h => (typeof h.agent === "string" ? h.agent : h.agent.name) === targetName)
    if (!declared) throw new Error(`agent "${this.name}" cannot hand off to "${targetName}"`)
    const resolver = this.runtimeBinding?.resolveAgent
    if (!resolver) throw new Error(`agent "${this.name}" requires a host target resolver`)
    const target = await resolver(targetName)
    if (!target) throw new Error(`target agent "${targetName}" is not registered`)
    const result = await target.run(request.goal, { metadata: request.metadata, providerOptions: request.providerOptions })
    return { output: result.output, status: result.status === "cancelled" ? "partial" : result.status }
  }
  async listen(): Promise<AgentRunResult | null> {
    const source = this.runtimeBinding?.signalSource
    if (!source) throw new Error("agent signals require runtimeBinding.signalSource")
    const claim = await source.claimSignal()
    if (!claim) return null
    const goal = String(claim.signal.payload.goal ?? claim.signal.payload.summary ?? JSON.stringify(claim.signal.payload))
    try { const result = await this.run(goal); await source.ackSignal(claim); return result }
    catch (error) { await source.nackSignal(claim); throw error }
  }
  async close() { await this.runtimeBinding?.mcpPlane?.disconnect() }
  async result(sessionId: string, events: StreamEvent[]): Promise<AgentRunResult> {
    const entries = await this.sessionLog.read(sessionId)
    const started = [...entries].reverse().find(e => e.event.kind === "run_started")?.event
    const terminal = [...entries].reverse().find(e => e.event.kind === "run_terminal")?.event
    const output = events.filter((e): e is StreamEvent & { delta: unknown } => e.type === "text_delta" && "delta" in e).map(e => String(e.delta)).join("")
    const usage = [...events].reverse().find((e): e is UsageEvent => e.type === "usage")
    const result: AgentRunResult = { output, runId: started?.kind === "run_started" ? started.run_id : undefined, sessionId, status: statusOf(terminal?.kind === "run_terminal" ? terminal.reason : events.some(e => e.type === "error") ? "error" : "completed") }
    const attempt = [...entries].reverse().find(e => e.event.kind === "provider_attempt")?.event
    if (usage) result.usage = { inputTokens: usage.inputTokens ?? 0, outputTokens: usage.outputTokens ?? 0, totalTokens: usage.totalTokens }
    else if (attempt?.kind === "provider_attempt" && attempt.usage) {
      result.usage = { inputTokens: attempt.usage.inputTokens, outputTokens: attempt.usage.outputTokens, totalTokens: attempt.usage.inputTokens + attempt.usage.outputTokens }
    }
    const measured = [...entries].reverse().find(e => e.event.kind === "prompt_measured")?.event
    const prepared = [...entries].reverse().find(e => e.event.kind === "context_prepared")?.event
    if ((started?.kind === "run_started" && started.route) || (attempt?.kind === "provider_attempt" && attempt.route) || measured || prepared) {
      result.evidence = {
        ...((attempt?.kind === "provider_attempt" ? attempt.route : started?.kind === "run_started" ? started.route : undefined) ? { route: attempt?.kind === "provider_attempt" ? attempt.route : (started as Extract<SessionEvent, { kind: "run_started" }>).route } : {}),
        ...(measured?.kind === "prompt_measured" ? { measurement: measured.measurement } : {}),
        ...(prepared?.kind === "context_prepared" ? { contextBinding: prepared.preparation.binding } : {}),
      }
    }
    if (this.outputSchema) {
      result.outputValidation = validateAgainstSchema(extractJsonValue(output), this.outputSchema)
      if (!result.outputValidation.ok && result.status === "completed") result.status = "failed"
    }
    return result
  }
}
